### PR TITLE
`Communication`: Add course wide search, filtering and sorting for chat messages

### PR DESCRIPTION
**Pull request by egekurt123**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

We currently have two communication modules: one named 'Communications,' which functions as a legacy system, and another named 'Messages.' We aim to consolidate them under the 'Messages' module and phase out the legacy 'Communications' module. To achieve this, implementing the missing course-wide search feature in the 'Messages' module is the first step.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with Communications enabled

1. Log in to Artemis
2. Navigate to Messages of a Course
3. Post messages to different channels
4. Search messages using the search bar over the sidebar
5. Make sure search functions correctly and everything else works as expected

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage 
|------------|--------------:|
| [course-conversations.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6356-27/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/course-conversations.component.ts.html) | 98.96% 
| [course-wide-search.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6356-27/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts.html) | 98.21%

### Screenshots

image

### Screencast and Explanation of How the Search Works

https://github.com/ls1intum/Artemis/assets/69898283/f427c44f-385c-4f56-b39b-73c7b92287cf

## Summary by CodeRabbit

- **New Features**
    - Added a search bar in the course conversations section to enhance user interaction with conversation threads.
    - Introduced a new course-wide search component to facilitate searching and interaction within course conversations.
- **Enhancements**
    - Updated conversation headers to display relevant information based on the user's current page.
    - Improved layout and visibility of conversation elements across different pages.
- **Bug Fixes**
    - Adjusted display logic in conversation messages to correctly reflect the active conversation context.
- **Documentation**
    - Enhanced in-app documentation accessibility through new documentation buttons.
- **Style**
    - Implemented new styles for the course-wide search functionality to improve user experience.

### CODE REVIEW

**Review by basak-akan — APPROVED** [review]
Tested on TS2, posting and searching messages worked well as described

File: src/main/webapp/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.ts [hunk 1]
```diff
@@ -68,6 +71,7 @@ export class ConversationMessagesComponent implements OnInit, AfterViewInit, OnD
     totalNumberOfPosts = 0;
     page = 1;
     public isFetchingPosts = true;
+    isallMessagesPage = false;
```
Comment (JohannesWt) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.ts:74]:
```suggestion
    isAllMessagesPage = false;
```

File: src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.ts [hunk 2]
```diff
@@ -349,7 +349,7 @@ export class ConversationSelectionSidebarComponent implements AfterViewInit, OnI
     }
 
     private setDisplayedChannels(channels: ChannelDTO[]): void {
-        this.displayedChannelConversations = channels;
+        this.displayedChannelConversations = channels.filter((channel) => channel.name !== 'all-messages');
```
Comment (JohannesWt) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.ts:352]: maybe it makes sense to define a constant for 'all-messages'

File: src/main/webapp/app/overview/course-conversations/course-conversations.component.html [hunk 3]
```diff
@@ -1,5 +1,18 @@
...
+            [(ngModel)]="searchText"
+            class="form-control"
+            (keyup.enter)="onSelectContext()"
+            placeholder="{{ 'artemisApp.metis.discussionSection.searchBarDefault' | artemisTranslate }}"
+        />
+        <jhi-button [btnType]="ButtonType.SECONDARY" [icon]="faTimes" (onClick)="searchText = ''; onSelectContext()" />
+        <jhi-button id="search-submit" class="ms-1" [icon]="faSearch" (onClick)="onSelectContext()" />
+        <jhi-documentation-button [type]="documentationType" class="information-button" />
+    </div>
     @if (course && isCodeOfConductAccepted === false) {
```
Comment (JohannesWt) [inline src/main/webapp/app/overview/course-conversations/course-conversations.component.html:16]: You deleted the comment - not sure if it is needed
```suggestion
<!-- only display after isCodeOfConductAccepted is loaded and set to false -->
    @if (course && isCodeOfConductAccepted === false) {
```

File: src/main/webapp/app/shared/metis/post/post.component.ts [hunk 4]
```diff
@@ -93,10 +93,8 @@ export class PostComponent extends PostingDirective<Post> implements OnInit, OnC
      * ensures that only when clicking on context without having control key pressed,
      * the modal is dismissed (closed and cleared)
      */
-    onNavigateToContext($event: MouseEvent) {
-        if (!$event.metaKey) {
-            this.modalRef?.dismiss();
-        }
+    onNavigateToContext() {
+        this.metisConversationService.setActiveConversation(this.contextInformation.queryParams!['conversationId']);
```
Comment (JohannesWt) [inline src/main/webapp/app/shared/metis/post/post.component.ts:97]: Why is the check if the control key is pressed not needed anymore?

**Review by JohannesWt — CHANGES_REQUESTED** [review]
see comments

**Comment by egekurt123** [comment]
Refreshing the page after searching show all messages and displays "Search results for ' ' "

**Review by florian-glombik — COMMENTED** [review]
I am not quite sure if it is a good idea to implement this with a client side search. I could imagine that a course accumulates quite some messages over the semester.
I am thinking of courses like Introduction to Software Engineering with more than 2000 students. At the end of the semester this could become super slow! 

Do we have numbers from previous courses that actively used the Messaging feature on how many messages will be accumulated (and would need to be filtered and sorted client side)?

**Comment by egekurt123** [comment]
> I am not quite sure if it is a good idea to implement this with a client side search. I could imagine that a course accumulates quite some messages over the semester. I am thinking of courses like Introduction to Software Engineering with more than 2000 students. At the end of the semester this could become super slow!
> 
> Do we have numbers from previous courses that actively used the Messaging feature on how many messages will be accumulated (and would need to be filtered and sorted client side)?

The search logic was already implemented in the legacy communications module, so there are existing queries that do this. I just reused the methods on the client side. In the ConversationMessageRepository class on the server side, the query is created.

**Review by JohannesWt — DISMISSED** [review]
Code looks good 👍 - Reapprove

**Review by basak-akan — DISMISSED** [review]
Reapproved, LGTM

**Review by coolchock — DISMISSED** [review]
re-approve

**Review by az108 — DISMISSED** [review]
Changes lgtm

**Review by edkaya — DISMISSED** [review]
Tested on ts1, works as expected!

**Comment by rabeatwork** [comment]
We should hide the search bar in the code of conduct page. 
image
image

**Comment by egekurt123** [comment]
> We should hide the search bar in the code of conduct page. image

fixed

**Review by sarpsahinalp — APPROVED** [review]
Reapprove

**Review by az108 — APPROVED** [review]
Reapprove

**Review by basak-akan — APPROVED** [review]
Reapprove

**Review by coolchock — APPROVED** [review]
re-approve

**Review by JohannesWt — APPROVED** [review]
Reapprove

**Pull request merged**